### DISCUSSION TITLE
Any documentation of WebP conversion defaults?

### DISCUSSION THREAD

**Discussion by selden** [ImageMagick/ImageMagick#7121] [Help]
### ImageMagick version

7.1.1-27 and 7.0.10-61

### Operating system, version and so on

Windows 10 Pro x64 22H2 build 19045.4046

### Description

Is there any documentation of the default parameter values used when converting to WebP?

For example, what values are applied when using the command 
convert image.png image.webp

My searches of GitHub and the wider Internet haven't been particularly helpful.

**Comment by snibgo** [comment]
The defaults seem to be set in the Webp library, module config_enc.c, function WebPConfigInitInternal(), not ImageMagick.

**Comment by mrienstra** [comment]
More details regarding defaults can be found here (as pointed out by @snibgo): [webmproject/libwebp src/webp/encode.h#L172-L179](https://github.com/webmproject/libwebp/blob/ac2795e/src/webp/encode.h#L172-L179) ➝ [webmproject/libwebp src/enc/config_enc.c#L27](https://github.com/webmproject/libwebp/blob/ac2795e/src/enc/config_enc.c#L27) _(links point to most recent commit on `main` as of 2025-08-06)_

Note that this is the latest code, to see which version of `libwebp` ImageMagick is using:

```
identify -list format | grep webp
```

Example output:

```
     WEBP* WEBP      rw+   WebP Image Format (libwebp 1.5.0 [0210])
```

Code for `v1.5.0`: [webmproject/libwebp src/webp/encode.h#L170-L177](https://github.com/webmproject/libwebp/blob/v1.5.0/src/webp/encode.h#L170-L177) ➝ [webmproject/libwebp src/enc/config_enc.c#L24](https://github.com/webmproject/libwebp/blob/v1.5.0/src/enc/config_enc.c#L24)

So defaults (for both `main` as of 2025-08-06, and `v1.5.0`) include _(there are more than this!)_:

- `quality = 75` (as per `encode.h`, all others as per `config_enc.c`)
- `alpha_filtering = 1` (`alpha-filtering` in ImageMagick)
- `alpha_quality = 100` (`alpha-quality` in ImageMagick)
- `autofilter = 0` (`auto-filter` in ImageMagick)
- `filter_sharpness = 0` (see preset values below, `filter-sharpness` in ImageMagick)
- `filter_strength = 60` (see preset values below, `filter-strength` in ImageMagick)
- `lossless = 0`
- `method = 4`
- `pass = 1` (see note below)
- `preprocessing = 0` (see preset values below)
- `segments = 4` (see preset values below)
- `sns_strength = 50` (see preset values below, `sns-strength` in ImageMagick)
- `target_PSNR = 0.` (`target-psnr` in ImageMagick)
- `target_size = 0` (`target-size` in ImageMagick)

## Presets

If you want to replicate a `libwebp` preset:

`WEBP_PRESET_PICTURE`: (digital picture, like portrait, inner shot)
```
-define webp:sns-strength=80 \
-define webp:filter-sharpness=4 \
-define webp:filter-strength=35 \
-define webp:preprocessing=0
```

`WEBP_PRESET_PHOTO`: (outdoor photograph, with natural lighting)
```
-define webp:sns-strength=80 \
-define webp:filter-sharpness=3 \
-define webp:filter-strength=30 \
-define webp:preprocessing=2
```

`WEBP_PRESET_DRAWING`: (hand or line drawing, with high-contrast details)
```
-define webp:sns-strength=25 \
-define webp:filter-sharpness=6 \
-define webp:filter-strength=10
```

`WEBP_PRESET_ICON`: (small-sized colorful images)
```
-define webp:sns-strength=0 \
-define webp:filter-strength=0 \
-define webp:preprocessing=0
```

`WEBP_PRESET_TEXT`: (text-like)
```
-define webp:sns-strength=0 \
-define webp:filter-strength=0 \
-define webp:preprocessing=0 \
-define webp:segments=2
```

## `pass`

https://developers.google.com/speed/webp/docs/cwebp#lossy_options says:
> `-pass int`
> Set a maximum number of passes to use during the dichotomy used by options `-size` or `-psnr`. Maximum value is `10`, default is `1`. If options `-size` or `-psnr` were used, but `-pass` wasn't specified, a default value of `6` passes will be used. If `-pass` is specified, but neither `-size` nor `-psnr` are, a target PSNR of 40dB will be used.

Looks like ImageMagick replicates some of this convenient behavior on your behalf, specifically it sets `pass=6` if `target-psnr` or `target-size` are specified -- see [coders/webp.c#L1205-L1207](https://github.com/ImageMagick/ImageMagick/blob/f86cbbd/coders/webp.c#L1205-L1207) (link points to most recent commit on `main` as of 2025-08-06). However it does not set `target_PSNR=40` if `pass` is specified without `target-psnr` or `target-size`.

## `quality`

If `quality` is specified, and `libwebp` is v1.0.0 or newer (v1.0.0 was released Apr 20, 2018), than ImageMagick sets `near_lossless` to the same value as `quality`, see [coders/webp.c#L1145-L1150](https://github.com/ImageMagick/ImageMagick/blob/f86cbbd/coders/webp.c#L1145-L1150). `near_lossless` cannot be set directly (independently of `quality`).

If the `quality` is set to `100` or greater, ImageMagick sets `lossless=1`, see [coders/webp.c#L1151-L1152](https://github.com/ImageMagick/ImageMagick/blob/f86cbbd/coders/webp.c#L1151-L1152). `lossless` is parsed after this, so you can force a lossy encode at quality 100+ if you wish: `-quality 100 -define webp:lossless=false`)

See also https://imagemagick.org/script/webp.php -- note that this shows `segment` instead of `segments`, which is not consistent with [the code](https://github.com/ImageMagick/ImageMagick/blob/f86cbbd/coders/webp.c#L1188).

**Comment by mrienstra** [comment]
PS: if anyone reading this would like to help improve the docs: 

https://github.com/ImageMagick/Website/blob/main/include/webp.php
or
https://github.com/ImageMagick/Website/issues/117 (if still relevant)

**Reply by urban-warrior** [reply]
Thanks for contributing.  Review the updated web page and let us know if further changes are needed.
